src/content/datachannel/datatransfer/js/main.js

```diff
@@ -41,97 +44,78 @@ megsToSend.addEventListener('change', function(e) {
 
   localConnection = new RTCPeerConnection(servers);
-  console.log('Created local peer connection object localConnection');
 
+  // Let's make a data channel!
   const dataChannelParams = {ordered: false};
   if (orderedCheckbox.checked) {
     dataChannelParams.ordered = true;
   }
-
   sendChannel = localConnection.createDataChannel('sendDataChannel', dataChannelParams);
   sendChannel.binaryType = 'arraybuffer';
-  console.log('Created send data channel');
+  sendChannel.bufferedAmountLowThreshold = CHUNK_SIZE / 2;
+  sendChannel.addEventListener('open', onSendChannelStateChange);
+  sendChannel.addEventListener('close', onSendChannelStateChange);
+  console.log('Created send data channel: ', sendChannel);
 
-  sendChannel.onopen = onSendChannelStateChange;
-  sendChannel.onclose = onSendChannelStateChange;
-  localConnection.onicecandidate = e => onIceCandidate(localConnection, e);
+  console.log('Created local peer connection object localConnection: ', localConnection);
 
-  localConnection.createOffer().then(gotDescription1, onCreateSessionDescriptionError);
+  localConnection.addEventListener('icecandidate', e => onIceCandidate(localConnection, e));
 
-  remoteConnection = remoteConnection = new RTCPeerConnection(servers);
-  console.log('Created remote peer connection object remoteConnection');
+  remoteConnection = new RTCPeerConnection(servers);
+  remoteConnection.addEventListener('icecandidate', e => onIceCandidate(remoteConnection, e));
+  remoteConnection.addEventListener('datachannel', receiveChannelCallback);
 
-  remoteConnection.onicecandidate = e => onIceCandidate(remoteConnection, e);
-  remoteConnection.ondatachannel = receiveChannelCallback;
-}
+  try {
+    const localOffer = await localConnection.createOffer();
+    await handleLocalDescription(localOffer);
+  } catch (e) {
+    console.error('Failed to create session description: ', e);
+  }
 
-function onCreateSessionDescriptionError(error) {
-  console.log(`Failed to create session description: ${error.toString()}`);
+  transferStatus.innerHTML = 'Peer connection setup complete.';
 }
 
-function randomAsciiString(length) {
-  let result = '';
-  for (let i = 0; i < length; i++) {
-    // Visible ASCII chars are between 33 and 126.
-    result += String.fromCharCode(33 + Math.random() * 93);
+function sendData(e) {
+  console.log('BufferedAmountLow event:', e);
+  if (sendProgress.value < sendProgress.max) {
+    transferStatus.innerHTML = 'Sending data...';
+    sendChannel.addEventListener('bufferedamountlow', sendData, {once: true});
+
+    // The following is a workaround due to the problem with bufferedamountlow event not being fired on every
+    // call to send(), despite the amount of data exceeding the bufferedAmountLowThreshold.
+    // If the event would be firing correctly, we could exclude the while loop below.
+    let count = 0;
+    while (sendProgress.value < sendProgress.max && sendChannel.bufferedAmount < CHUNK_SIZE) {
```

Comment:
We don't want too many calls to `send()` as that will block the main thread and stall the UI. What we could do is to figure out the largest possible chunk and then just send that and wait for `bufferedlowamount` to trigger. 

I also don't want to add a workaround like `setTimeout()` for this, as we decided on not doing that earlier.

On a higher level, this sample should probably be removed and replaced with something more relevant (the filetransfer is one example that works better). Currently we run into weird scenarios because of a very unrealistic use of the API.

Reply:
`setTimeout` works differently here. This is not about polling, this is about not racing with the underlying transport as that will send in parallel and `bufferedAmount` reflects the live value. In the worst case, this can lead to blocking the UI for the whole send operation (e.g. it will effectively send all chunks at once in a blocking loop).

I agree that the example is artificial but the required send mechanism would not change much.

Reply:
Do you mind making a PR with the suggested change? I'm not entirely sure how you would like this too look like and I feel it is better if I review the change instead of trying to guess what you want. ;)

Reply:
Probably makes sense.